### PR TITLE
DLPack Conversion API

### PR DESCRIPTION
**Pull request by eqy**
(do not merge yet)
cc @tqchen

### CODE REVIEW

File: python/tvm/_ffi/ndarray.py [hunk 1]
```diff
@@ -126,6 +126,17 @@ def empty(shape, dtype="float32", ctx=context(1, 0)):
 
 
 def from_dlpack(dltensor):
+    """Produce an array from a DLPack tensor without memory copy
+
+    Parameters
+    ----------
+    dltensor : DLPack tensor
+
```
Comment (tqchen) [inline python/tvm/_ffi/ndarray.py]: add description

File: python/tvm/contrib/dlpack.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+from .. import ndarray
+
+def convert_func(tvm_func, tensor_type, to_dlpack_func):
+    """Convert a tvm function into one that accepts a tensor from another
+       framework, provided the other framework supports DLPACK
+
+    Parameters
+    ----------
+    tvm_func: Built tvm function operating on arrays
```
Comment (tqchen) [inline python/tvm/contrib/dlpack.py]: numpy docstring convention
```python
tvm_func: Function
     The original tvm function input
````

File: include/tvm/runtime/c_runtime_api.h [hunk 3]
```diff
@@ -445,6 +445,29 @@ TVM_DLL int TVMArrayCopyFromTo(TVMArrayHandle from,
...
+ * \brief Produce a DLMangedTensor from the array that shares data memory with
+ * the array.
+ * \param from The source array.
+ * \param out The DLManagedTensor handle.
+ * \return 0 when success, -1 when failure happens
+ */
+TVM_DLL int TVMArrayToDLPack(TVMArrayHandle from,
+                             DLManagedTensor** out);
+
+/**/
```
Comment (tqchen) [inline include/tvm/runtime/c_runtime_api.h]: Add comment block here

File: python/tvm/contrib/dlpack.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    tvm_func: Function
+        Built tvm function operating on arrays
+
+    Returns
+    -------
+    wrapped_func: Function
+        Wrapped tvm function that operates on PyTorch tensors
+    """
+    import torch
+    import torch.utils.dlpack
```
Comment (tqchen) [inline python/tvm/contrib/dlpack.py]: add a check to see if tvm_func is callable and raise error if it is not

File: python/tvm/contrib/dlpack.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    to_dlpack_func: Function
+        Function to convert the source tensors to DLPACK
+    """
+    def _wrapper(*args):
+        args = tuple(ndarray.from_dlpack(to_dlpack_func(arg))\
+            if isinstance(arg, tensor_type) else arg for arg in args)
+        return tvm_func(*args)
+    return _wrapper
+
+def to_pytorch(tvm_func):
```
Comment (tqchen) [inline python/tvm/contrib/dlpack.py]: to_pytorch_func to avoid confusion

**Comment by tqchen** [comment]
Thanks @eqy ! this is merged

**Comment by tqchen** [comment]
@wkcn Thanks for pointing this out, that line should be changed to     
```pycapsule = ctypes.cast(pycapsule, ctypes.py_object)```, can you send a PR?

**Comment by tqchen** [comment]
cc @eqy

**Comment by tqchen** [comment]
https://github.com/dmlc/tvm/pull/1589 should fix the problem

**Comment by wkcn** [comment]
Yes. #1589 has fixed the problem :-)

**Pull request merged**